src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java

```diff
@@ -155,5 +162,61 @@ public static List<ClinicalDataCount> normalizeDataCounts(List<ClinicalDataCount
         return new ArrayList<>(normalizedDataCounts);
     }
     
-    
+    public static void mergeDataFilterNumericalValues(StudyViewFilter studyViewFilter) {
+        if (studyViewFilter.getGenomicDataFilters() != null && !studyViewFilter.getGenomicDataFilters().isEmpty()) {
+            List<GenomicDataFilter> mergedGenomicDataFilters = mergeDataFilters(studyViewFilter.getGenomicDataFilters());
+            studyViewFilter.setGenomicDataFilters(mergedGenomicDataFilters);
+        }
+
+        if (studyViewFilter.getClinicalDataFilters() != null && !studyViewFilter.getClinicalDataFilters().isEmpty()) {
+            List<ClinicalDataFilter> mergedClinicalDataFilters = mergeDataFilters(studyViewFilter.getClinicalDataFilters());
+            studyViewFilter.setClinicalDataFilters(mergedClinicalDataFilters);
+        }
+
+        if (studyViewFilter.getGenericAssayDataFilters() != null && !studyViewFilter.getGenericAssayDataFilters().isEmpty()) {
+            List<GenericAssayDataFilter> mergedGenericAssayDataFilters = mergeDataFilters(studyViewFilter.getGenericAssayDataFilters());
+            studyViewFilter.setGenericAssayDataFilters(mergedGenericAssayDataFilters);
+        }
+    }
+
+    /**
+     * Merge the range of numerical bins in DataFilters to reduce the number of scans that runs on the database when filtering.
+     */
+    private static <T extends DataFilter> List<T> mergeDataFilters(List<T> filters) {
+        List<T> mergedDataFilters = new ArrayList<>();
+
+        for (T filter : filters) {
+            List<DataFilterValue> mergedValues = new ArrayList<>();
+
+            BigDecimal mergedStart = null, mergedEnd = null;
+            for (DataFilterValue dataFilterValue : filter.getValues()) {
```

Comment:
this logic only works if the bins are sorted properly, right? are we sure that bins are always sorted the way we want?

Reply:
I had 2 assumptions: 1) they come in sorted and 2) they're continuous. I handled the discontinuous case by adjacent merging but if sorting is needed we'll need an extra part and way more combinations of possible cases. What would you suggest?